models/resnet/resnet.hpp

```diff
@@ -0,0 +1,554 @@
+  //! Get Layers of the model.
+  ann::FFN<OutputLayerType, InitializationRuleType>& GetModel()
+  {
+    return resNet;
+  }
+
+  //! Load weights into the model and assumes the internal matrix to be
+  //  named "ResNet".
+  void LoadModel(const std::string& filePath);
+
+  //! Save weights for the model and assumes the internal matrix to be
+  //  named "ResNet".
+  void SaveModel(const std::string& filepath);
+
+ private:
+  /**
+   * Adds a 3x3 Convolution Block.
+   *
+   * @tparam SequentialType Layer type in which 3x3 convolution block will
+   *     be added.
+   *
+   * @param inSize Number of input maps.
+   * @param outSize Number of output maps.
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   * @param kernelWidth Width of the filter/kernel.
+   * @param kernelHeight Height of the filter/kernel.
+   * @param padW Padding width of the input.
+   * @param padH Padding height of the input.
+   */
+  template<typename SequentialType = ann::Sequential<>>
+  void ConvolutionBlock3x3(SequentialType* baseLayer,
+                           const size_t inSize,
+                           const size_t outSize,
+                           const size_t strideWidth = 1,
+                           const size_t strideHeight = 1,
+                           const size_t kernelWidth = 3,
+                           const size_t kernelHeight = 3,
+                           const size_t padW = 1,
+                           const size_t padH = 1)
+  {
+    baseLayer->Add(new ann::Convolution<>(inSize, outSize, kernelWidth,
+        kernelHeight, strideWidth, strideHeight, padW, padH, inputWidth,
+        inputHeight));
+
+    mlpack::Log::Info << "Convolution: " << "(" << inSize << ", " << inputWidth
+        << ", " << inputHeight << ")" << " ---> (";
+
+    // Updating input dimensions.
+    inputWidth = ConvOutSize(inputWidth, kernelWidth, strideWidth, padW);
+    inputHeight = ConvOutSize(inputHeight, kernelHeight, strideHeight, padH);
+
+    mlpack::Log::Info << outSize << ", " << inputWidth << ", " << inputHeight
+        << ")" << std::endl;
+  }
+
+  /**
+   * Adds a 1x1 Convolution Block.
+   *
+   * @tparam SequentialType Layer type in which 1x1 convolution block will
+   *     be added.
+   *
+   * @param inSize Number of input maps.
+   * @param outSize Number of output maps.
+   * @param downSampleInputWidth Input widht for downSample block.
+   * @param downSampleInputHeight Input height for downSample block.
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   * @param kernelWidth Width of the filter/kernel.
+   * @param kernelHeight Height of the filter/kernel.
+   * @param padW Padding width of the input.
+   * @param padH Padding height of the input.
+   * @param downSample Bool if it's a downsample block or not. Default is false.
```

Comment:
Conv1x1 and Conv3x3 can be combined into a single function. Seems repetitive, passing correct params should help.

Reply:
Hey, they are seperate because just having these two with default params saves a lot of other code and supplying values everywhere. I would like to keep it this way if that's possible.

Reply:
combining them complicates the whole code a lot more and is basically non understable then.

Reply:
Lets do one thing, share the snippet you wrote that you feel is complicated we can see if its possible to simplify that. As general practice, follow : DRY principle.

Reply:
Yup i get that but removing duplicate code here would make things complex and i don't have some code to show your right now but i can explain why i did this in the first place. 

1. So for now the conv1x1 has two functions or ways where it uses different image size depending if it's a down sample block or not.
2. If we merge both the blocks this would still need to be set somewhere inside the function with some kind of condition  
3.  if we merge them I would need to specify all the parameters at every place i use a 3x3 block or a 1x1 block, I can make it look clean without merging them and specifying only the params that i need. 
4. and you can see in the code for basicblock and bottleneck block, I have only specified stride width and stride height for the first layer and second layer respectively, if i merge them I would again need some condition and would need to pass all the params again.

I guess these were some of the things that came in my mind when i thought about merging these two block initially. it's not that i don't want to make the code more concise but i would like to avoid making it more complicated. and keeping them sepearte allows greater level of freedom and simplicity.

Reply:
Do you want me to write it and see if its simple?

Reply:
Yeah it will be great, also i do want to say that i am not trying to avoid the change. :smile: 
btw i am pushing some of the requested changes maybe then you can base your code upon that ?

Reply:
Hmm, Will share it, For mobile net lets spend more time on what should be coded than just making it work.

Reply:
Yup Sure, Do you want to discuss that over a call?  that way it would be more clear to me.

Reply:
```
void ConvolutionBlock(SequentialType* baseLayer,
                   	  const size_t inSize,
                       	   const size_t outSize,
                            const size_t strideWidth,
                            const size_t strideHeight,
                            const size_t kernelWidth,
                            const size_t kernelHeight,
                            const size_t padW = 0,
                            const size_t padH = 0,
                            const bool downSample = false,
                            const size_t downSampleInputWidth = 0,
                            const size_t downSampleInputHeight = 0)
{

	if (downsample)
	{
		inputWidth = downSampleInputWidth;
		inputHeigt = downSampleInputHeight;
	}

	baseLayer->Add(new ann::Convolution<>(inSize, outSize, kernelWidth,
           kernelHeight, strideWidth, strideHeight, padW, padH,
           inputWidth, inputHeight));
	baseLayer->Add(new ann::BatchNorm<>(outSize, 1e-5));
	// Updating input dimensions.
	inputWidth = ConvOutSize(inputWidth, kernelWidth, strideWidth, padW);
	inputHeight = ConvOutSize(inputHeight, kernelHeight, strideHeight, padH);	
}
```
If correctly parameters are passed this is all that is required to implement all blocks. This reduced about 50 or so lines or redundant code. Always a single function would be called with different params.

Reply:
> Yup Sure, Do you want to discuss that over a call? that way it would be more clear to me.

Sure, You can form an initial draft of the class and then we can discuss it.

Reply:
Hi, I have applied these suggestions, can you review the PR once again ?